tests/rptest/services/kerberos.py

```diff
@@ -468,28 +489,54 @@ def start_node(self, node, **kwargs):
-            principal=principal, keytab_file=self.keytab_file),
-                                           realm=self.kdc.realm,
-                                           principal=self.kdc.kadmin_principal,
-                                           password=self.kdc.kadmin_password,
-                                           krb5_conf_path=self.krb5_conf_path)
-        self.logger.debug(f"ktadd command: {cmd}")
-        node.account.ssh(f"mkdir -p {os.path.dirname(self.keytab_file)}")
-        node.account.ssh(cmd=cmd, allow_fail=False)
+        self.kdc.configure_service_principal(self.krb5_conf_path, principal,
+                                             self.keytab_file, node)
         super().start_node(node, **kwargs)
 
     def _service_principal(self, node, primary: str = "redpanda"):
         fqdn = RedpandaService.get_node_fqdn(node)
         return f"{primary}/{fqdn}@{self.realm}"
+
+
+class ActiveDirectoryKdc:
+    def __init__(self, logger, remote_account: RemoteAccount, realm: str,
+                 keytab_password: str, upn_user: str, spn_user: str):
+        self.logger = logger
+        self.remote_account = remote_account
+        self.realm = realm
+        self.keytab_password = keytab_password
+        self.upn_user = upn_user
+        self.spn_user = spn_user
+        self.nodes = [
+            namedtuple("node", "name account")(remote_account.hostname,
+                                               remote_account)
+        ]
+
+    def _configure_principal(self, krb5_conf_path: str, user: str,
+                             principal: str, password: str, dest: str,
+                             dest_node):
+        if not "@" in principal:
+            principal = f"{principal}@{self.realm}"
+        src = fr"{user}.keytab.temp"
+        cmd = f"ktpass -out {src} -mapuser {user} -princ {principal} -crypto ALL -pass {password} -ptype KRB5_NT_PRINCIPAL +DumpSalt"
+        self.logger.debug(f"Configuring principal cmd: {cmd}")
+        out = self.remote_account.ssh_output(cmd=cmd, allow_fail=False)
+        self.logger.info(f"Output of ktpass: {out}")
+        self.logger.debug(
+            f"copying keytab: from {src} to {dest_node.name} at {dest}")
+        dest_node.account.ssh(f"mkdir -p {os.path.dirname(dest)}")
+        self.remote_account.copy_between(src, dest, dest_node)
+        self.remote_account.ssh(cmd=f"del {src}", allow_fail=False)
+
+    def configure_service_principal(self, krb5_conf_path: str, principal: str,
+                                    dest: str, dest_node):
+        self._configure_principal(krb5_conf_path, self.spn_user, principal,
+                                  self.keytab_password, dest, dest_node)
+
+    def configure_user_principal(self, krb5_conf_path: str, principal: str,
+                                 dest: str, dest_node):
+        self._configure_principal(krb5_conf_path, self.upn_user, principal,
+                                  self.keytab_password, dest, dest_node)
```

Comment:
Like a dirty way of doing inheritance

Reply:
I could quite easily create a base class and name the type. Duck typing works, too.